### PR TITLE
feat(tenant): add subtenant concept; improve tenant parsing performance

### PR DESCRIPTION
**Pull request by juliusmh**
Adds the concepts of `subtenants` which is expressed through org IDs that have the form `<tenantID>:<subtenantID>`. Functions `TenantID` and `TenantIDs` will omit the subtenant part for compatibility. A new function `SubtenantID` extracts tenant and subtenant information.

`TenantID(..)` is used in many hot paths. Adding the subtenant ID logic had some negative performance impacts, hence I took some optimization measures:

- `ValidTenantID` uses a lookup table. This allows the function to be inlined. 
- Use `strings.IndexByte` instead of `strings.Contains` where possible. The former has SIMD optimized implementations.
- `TenantID` and `SubtenantID` have zero allocations, due to improved duplicate detection.

**BenchmarkTenantId:**
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/dskit/tenant
cpu: Apple M4 Pro
                   │ /tmp/tenant_bench_old │        /tmp/tenant_bench_new2         │
                   │        sec/op         │   sec/op     vs base                  │
TenantID/single-14            10.420n ± 6%   9.552n ± 5%   -8.33% (p=0.005 n=10+6)
TenantID/multi-14              60.24n ± 1%   14.72n ± 1%  -75.56% (p=0.000 n=10+6)
geomean                        25.05n        11.86n       -52.67%

                   │ /tmp/tenant_bench_old │          /tmp/tenant_bench_new2           │
                   │         B/op          │    B/op     vs base                       │
TenantID/single-14            0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10+6) ¹
TenantID/multi-14             48.00 ± 0%      0.00 ± 0%  -100.00% (n=10+6)
geomean                                  ²               ?                         ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean

                   │ /tmp/tenant_bench_old │          /tmp/tenant_bench_new2           │
                   │       allocs/op       │ allocs/op   vs base                       │
TenantID/single-14            0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10+6) ¹
TenantID/multi-14             1.000 ± 0%     0.000 ± 0%  -100.00% (n=10+6)
geomean                                  ²               ?                         ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean

```

**BenchmarkSubtenantID:**
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/dskit/tenant
cpu: Apple M4 Pro
BenchmarkSubtenantID/single-with-subtenant-14         	107781979	        11.06 ns/op	       0 B/op	       0 allocs/op
BenchmarkSubtenantID/single-no-subtenant-14           	100000000	        10.78 ns/op	       0 B/op	       0 allocs/op
BenchmarkSubtenantID/multi-same-tenant-14             	86582684	        14.35 ns/op	       0 B/op	       0 allocs/op
```

**BenchmarkStringsCut:**
```goos: darwin
goarch: arm64
pkg: github.com/grafana/dskit/tenant
cpu: Apple M4 Pro
BenchmarkStringsCut/strings.Cut/short-no-match-14         	405862113	         2.809 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/stringsCut/short-no-match-14          	564547033	         2.092 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/strings.Cut/short-match-14            	449214154	         2.690 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/stringsCut/short-match-14             	581317662	         2.077 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/strings.Cut/medium-no-match-14        	413275023	         2.913 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/stringsCut/medium-no-match-14         	549032720	         2.190 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/strings.Cut/medium-match-end-14       	406564318	         2.948 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/stringsCut/medium-match-end-14        	507624324	         2.368 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/strings.Cut/long-no-match-14          	411175990	         2.934 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/stringsCut/long-no-match-14           	545224579	         2.200 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/strings.Cut/long-match-end-14         	406851376	         2.946 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringsCut/stringsCut/long-match-end-14          	520922545	         2.311 ns/op	       0 B/op	       0 allocs/op
PASS
```

---

> [!NOTE]
> **Medium Risk**
> Changes core tenant/orgID parsing and expands the public `Resolver` interface; behavior changes and new validation paths could affect authentication/authorization-like request scoping if downstream assumptions differ.
> 
> **Overview**
> Adds **subtenant support** by allowing org IDs of the form `tenantID:subtenantID`, while keeping `TenantID`/`TenantIDs` backward-compatible by stripping the `:subtenant` portion.
> 
> Introduces `SubtenantID` and `SubtenantIDs` APIs (and wires them into `Resolver`/`MultiResolver`) to extract and validate subtenant information, and expands test coverage for mixed tenant/subtenant and multi-tenant edge cases.
> 
> Reworks tenant parsing/validation for performance: switches separators to bytes, adds a precomputed valid-character lookup table for `ValidTenantID`, and replaces higher-level string ops with `strings.IndexByte`-based helpers to reduce allocations in hot paths.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 17c8a6b20127e026e5a7dd69619713ecfded8928. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: tenant/resolver.go [hunk 1]
```diff
@@ -21,28 +27,51 @@ func TenantID(ctx context.Context) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	if !strings.Contains(orgID, tenantIDsSeparator) {
-		if err := ValidTenantID(orgID); err != nil {
+
+	// Fast path: no multi-tenant separator
+	idx := strings.IndexByte(orgID, tenantIDsSeparator)
```
Comment (tcard) [inline tenant/resolver.go]: **suggestion:** I find `strings.Cut` to be easier to work with generally than index-based finding/slicing.

File: tenant/resolver.go [hunk 2]
```diff
@@ -52,21 +81,109 @@ func TenantIDs(ctx context.Context) ([]string, error) {
...
 	}
-
 	return NormalizeTenantIDs(orgIDs), nil
 }
 
+// SubtenantID returns the subtenant ID from the context, or an empty string if
+// no subtenant is present. The orgID format is "tenantID:subtenantID" (e.g., "123456:k6").
+//
+//nolint:revive
+func SubtenantID(ctx context.Context) (string, string, error) {
```
Comment (tcard) [inline tenant/resolver.go]: We could make this more obvious:

```suggestion
func SubtenantID(ctx context.Context) (tenantID string, subtenantID string, err error) {
```

**Review by tcard — APPROVED** [review]

File: tenant/resolver.go (hunk 1 above)
Comment (juliusmh) [reply tenant/resolver.go]: Added `stringsCut` function, with benchmark to explain why.

File: tenant/resolver.go [hunk 3]
```diff
@@ -23,14 +23,14 @@ import (
...
+	orgIDs, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		return "", err
 	}
 
 	// Fast path: no multi-tenant separator
-	idx := strings.IndexByte(orgID, tenantIDsSeparator)
-	if idx == -1 {
+	orgID, remaining, ok := stringsCut(orgIDs, tenantIDsSeparator)
+	if !ok {
```
Comment (tcard) [inline tenant/resolver.go]: **suggestion:** Rename `ok` to `hasMoreIDs`, remove this if, do `for hasMoreIDs` instead. Less duplicate code this way.

Same for `SubtenantID`.

File: tenant/resolver.go [hunk 4]
```diff
@@ -12,37 +12,57 @@ import (
...
+	orgIDs, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		return "", err
 	}
-	if !strings.Contains(orgID, tenantIDsSeparator) {
-		if err := ValidTenantID(orgID); err != nil {
+
+	// Fast path: no multi-tenant separator
+	orgID, remaining, hasMoreIDs := stringsCut(orgIDs, tenantIDsSeparator)
+	if !hasMoreIDs {
```
Comment (tcard) [inline tenant/resolver.go]: Sorry, my main point with the previous suggestion was that this whole `if` block is redundant if the `for` already checks `hasMoreIDs`. The early return doesn't buy us anything here (I think).

Comment (juliusmh) [reply tenant/resolver.go]: Ah yeah, good point. I consolidated the fast/slow paths

File: tenant/resolver.go [hunk 5]
```diff
@@ -52,21 +72,99 @@ func TenantIDs(ctx context.Context) ([]string, error) {
...
 	}
-
 	return NormalizeTenantIDs(orgIDs), nil
 }
 
+// SubtenantID returns the subtenant ID from the context, or an empty string if
+// no subtenant is present. The orgID format is "tenantID:subtenantID" (e.g., "123456:k6").
+//
+//nolint:revive
+func SubtenantID(ctx context.Context) (subtenantID string, tenantID string, _ error) {
```
[the remark this answers is not in the document]
Comment (juliusmh) [reply tenant/resolver.go]: Fixed.
